ui/src/saros/ui/browser/SWTBrowserWrapper.java

```diff
@@ -0,0 +1,192 @@
+package saros.ui.browser;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import org.eclipse.swt.browser.Browser;
+import org.eclipse.swt.browser.BrowserFunction;
+import org.eclipse.swt.browser.ProgressEvent;
+import org.eclipse.swt.browser.ProgressListener;
+import org.eclipse.swt.events.DisposeEvent;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Display;
+
+public class SWTBrowserWrapper implements IBrowserWrapper {
+
+  private enum ReadyState {
+    LOADING,
+    COMPLETE
+  }
+
+  private final Browser browser;
+  private ReadyState state = ReadyState.LOADING;
+  private List<Runnable> runOnDisposalList = new ArrayList<Runnable>();
+  private final ProgressListener readyStateListener =
+      new ProgressListener() {
+        @Override
+        public void completed(ProgressEvent event) {
+          state = ReadyState.COMPLETE;
+        }
+
+        @Override
+        public void changed(ProgressEvent event) {
+          state = ReadyState.LOADING;
+        }
+      };
+
+  public SWTBrowserWrapper(Composite parent, int style) {
+    browser = new Browser(parent, style);
+
+    browser.addDisposeListener(
+        (DisposeEvent e) -> {
+          for (Runnable runnable : runOnDisposalList) {
+            runnable.run();
+          }
+        });
+  }
+
+  @Override
+  public boolean setFocus() {
+    return this.syncExec(
+        () -> {
+          return browser.setFocus();
+        });
+  }
+
+  @Override
+  public void setSize(final int width, final int height) {
+    this.syncExec(() -> browser.setSize(width, height));
+  }
+
+  @Override
+  public String getUrl() {
+    return this.syncExec(
+        () -> {
+          return browser.getUrl();
+        });
+  }
+
+  @Override
+  public boolean loadUrl(final String url, final long timeout) {
+    browser.getShell().open();
+
+    this.state = ReadyState.LOADING;
+    browser.addProgressListener(readyStateListener);
+
+    browser.setUrl(url);
+    boolean successful = this.waitForPageLoaded(url, timeout);
+
+    browser.removeProgressListener(readyStateListener);
+
+    return successful;
+  }
+
+  /**
+   * Waits for a page to be loaded.
+   *
+   * @param url
+   * @param timeout
+   * @return
+   */
+  private boolean waitForPageLoaded(final String url, final long timeout) {
+
+    Runnable waitForLoaded =
+        () -> {
+          while (this.state != ReadyState.COMPLETE) {
+            if (!browser.getDisplay().readAndDispatch()) {
+              browser.getDisplay().sleep();
+            }
+          }
+        };
+
+    Thread t = new Thread(waitForLoaded);
+    t.setDaemon(true);
+    t.start();
+
+    try {
+      t.join(timeout);
+    } catch (InterruptedException e) {
+      return false;
+    }
+    return true;
```

Comment:
This method always returns true unless the caller somehow interrupts the thread that his method was calling.

Reply:
I will improve this so that the method returns only **true** if the loading process was successful.